gadget/gadget.go

```diff
@@ -816,6 +817,24 @@ func validateVolume(vol *Volume, knownFsLabelsPerVolume map[string]map[string]bo
 		var start quantity.Offset
 		if s.Offset != nil {
 			start = *s.Offset
+
+			// The GPT partition table will start from sector 2 with length
+			// 16384 bytes
+			if vol.Schema == schemaGPT {
+				if start < 17408 {
```

Comment:
Sorry, actually there would be an overlap between a partition with offset `start` and size `size` and the GPT table if
```
start <= GPT_start + GPT_size && start + size > GPT_start
```
If we take as GPT header the union of GPTs for 512 and 4096 (worst case for initial and final offsets), we would have
```
start <= 4096*34 && start + size > 512*2
```

Reply:
If I understand correctly, the sector 1 is for GPT header, so I assume we should not overwrite it as well, I am not sure whether there is any critical data in the sector 0 or not, but I guess it might not make sense for user to put something inside the sector 0 only?

Reply:
You are right, so in fact the condition is:
```
start <= 4096*34 && start + size > 512
```
Note that actually we do write in sector 0 in the case of pc-gadget: https://github.com/snapcore/pc-amd64-gadget/blob/20/gadget.yaml#L6 - This is an mbr for legacy BIOS support (actually an important case for cloud).

Reply:
I took a look at the UEFI spec in the end, I was sort of assuming that the usual size for the partition entries was always 32 LBAs but indeed that is not necessarily the case. Quoting from the spec so we keep track of this:

"If the block size is 512, the First Usable LBA must be greater than or equal to 34 (allowing 1 block for the Protective
MBR, 1 block for the Partition Table Header, and 32 blocks for the GPT Partition Entry Array); if the logical block
size is 4096, the First Useable LBA must be greater than or equal to 6 (allowing 1 block for the Protective MBR, 1
block for the GPT Header, and 4 blocks for the GPT Partition Entry Array)."

I think it is fair to assume that usually we will get the minimum as of six 4KiB sector in real life as you have done.

In my last comments I moved to thinking that maybe it would be better to move to print a warning for all possible sector sizes to avoid overcomplicating the condition. Assuming this and with the new numbers, we would print a warning if
```
start <= 4096*6 && start + size > 512
```
(which checks if a chunk of data starting at "start" and with "size" overlaps with GPT data for any block size - this won't happen if the data lives in sector 0 or in higher addresses that the end of the GPT data).

Otherwise, if you prefer to do different things for 512 vs 4KiB
```
if (start <= 512*34 && start + size > 512)
    return error
else if (start <= 4096*6 && start + size > 4096)
    print warning
```
(there are probably equivalent ways of doing these checks anyway)

Reply:
Well, actually, we do write in the first sector: we write the MBR in some case, see https://github.com/snapcore/pc-amd64-gadget/blob/20/gadget.yaml#L6 (offset is implicitly 0 there).

Reply:
Ah, looks like I misunderstood this, then I will follow the way you suggested and push the commit after tested

Reply:
Hi, 

for the offset between 512 and 4096, it seems that we should not simply return error, because it should be allowed to write data between 512 and 4096 if sector size is 4096, so I think it might be better if we do this
```go
if (start < 512 * 34 && start + size >= 4096)
    return error
else if (start < 4096 && start + size >= 512) || (start < 4096 * 6 && start + size >= 512 * 34)
    print warning
```

Reply:
Hi Aristo,

The ranges seems fine, but the second condition can be simplified. As you are saying:
```
if data intersects (4096, 512*34)
    return error
else if data intersects (512, 4096) OR (512*34, 4096*6)
    print warning
```
The second is an else if, we already checked the segment in the "if" part so it is fine if it is included in the else if:
```
if data intersects (4096, 512*34)
    return error
else if data intersects (512, 4096*6)
    print warning
```
So it should be
```
if (start <= 512*34 && start + size > 4096)
    return error
else if (start <= 4096*6 && start + size > 512)
    print warning
```
Note also that you had the `=` in the wrong part of the comparison (think of case start=0 and size=4096, which should not return an error).

Reply:
Sounds good! I think that we finally converged :slightly_smiling_face: .